File: app/src/main/java/com/theandroiddev/riffking/HomeActivity.java
package com.theandroiddev.riffking;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.mikhaellopez.circularimageview.CircularImageView;
import com.squareup.picasso.Picasso;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class HomeActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener,
        HomeFragment.OnFragmentInteractionListener, RankingFragment.OnFragmentInteractionListener,
        VideosFragment.OnFragmentInteractionListener, UploadFragment.OnFragmentInteractionListener,
        ThreadFragment.OnFragmentInteractionListener, InsertThreadFragment.OnFragmentInteractionListener,
        ProfileFragment.OnFragmentInteractionListener, ProfileCommentsFragment.OnFragmentInteractionListener,
        ProfileVideosFragment.OnFragmentInteractionListener, ProfileRepFragment.OnFragmentInteractionListener,
        ProfileRepFragmentMe.OnFragmentInteractionListener {

    private static final String TAG = "HomeActivity";
    static User user;
    static Thread thread;
    public FloatingActionButton fab;
    public CircularImageView navProfileImg;
    public TextView navNameTv, navEmailTv;
    DrawerLayout drawer;
    String urlLink = "";
    FirebaseAuth mFirebaseAuth;
    FirebaseAuth.AuthStateListener mAuthStateListener;
    private DatabaseReference mDatabase;

    public static List<Thread> loadSharedPreferencesLogList(Context context) {
        List<Thread> callLog = new ArrayList<Thread>();
        SharedPreferences mPrefs = context.getSharedPreferences("threadsSaved", Context.MODE_PRIVATE);
        Gson gson = new Gson();
        String json = mPrefs.getString("myJson", "");
        if (json.isEmpty()) {
            callLog = new ArrayList<Thread>();
        } else {
            Type type = new TypeToken<List<Thread>>() {
            }.getType();
            callLog = gson.fromJson(json, type);
        }
        return callLog;
    }

    public static void saveSharedPreferencesLogList(Context context, List<Thread> callLog) {
        SharedPreferences mPrefs = context.getSharedPreferences("threadsSaved", Context.MODE_PRIVATE);
        SharedPreferences.Editor prefsEditor = mPrefs.edit();
        Gson gson = new Gson();
        String json = gson.toJson(callLog);
        prefsEditor.putString("myJson", json);
        prefsEditor.commit();
    }

    @Override
    protected void onResume() {
        super.onResume();

    }

    @Override
    protected void onStart() {
        super.onStart();
        mFirebaseAuth.addAuthStateListener(mAuthStateListener);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);

        mDatabase = FirebaseDatabase.getInstance().getReference();

        mFirebaseAuth = FirebaseAuth.getInstance();

        initUser();

        mDatabase.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                if (dataSnapshot.child("users").child(user.getId()).getValue(User.class) == null) {
                    logout();
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });

        mAuthStateListener = new FirebaseAuth.AuthStateListener() {
            @Override
            public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
                if (firebaseAuth.getCurrentUser() == null) {
                    startActivity(new Intent(HomeActivity.this, LoginActivity.class));
                }
            }
        };


        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        fab = (FloatingActionButton) findViewById(R.id.fab);

        drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.setDrawerListener(toggle);
        toggle.syncState();

        final NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
        final View headerView = navigationView.getHeaderView(0);
        navProfileImg = (CircularImageView) headerView.findViewById(R.id.nav_profile_img);
        navProfileImg.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                openProfileFragment(headerView);
                drawer.closeDrawer(GravityCompat.START);
            }
        });
        navNameTv = (TextView) headerView.findViewById(R.id.nav_name_tv);
        navEmailTv = (TextView) headerView.findViewById(R.id.nav_email_tv);

        setName();


        HomeFragment fragment = new HomeFragment();
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.content_home, fragment).commit();

        youtubeShareable();

    }

//    @Override
//    public boolean onKeyDown(int keyCode, KeyEvent event) {
//        if (keyCode == KeyEvent.KEYCODE_BACK) {
//            onBackPressed();
//        }
//        return super.onKeyDown(keyCode, event);
//    }
//
//    @Override
//    public void onBackPressed() {
//        Log.d(TAG, "onBackPressed: ");
//
//        if (drawer.isDrawerOpen(GravityCompat.START)) {
//            drawer.closeDrawer(GravityCompat.START);
//
//        } else {
//            if (getFragmentManager().getBackStackEntryCount() > 1) {
//                getFragmentManager().popBackStack();
//            } else {
//                super.onBackPressed();
//            }
//        }
//
//    }

    private void youtubeShareable() {
        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            urlLink = extras.getString(Intent.EXTRA_TEXT);

            if (urlLink != null && !urlLink.equals("")) {
                displaySelectedScreen(R.id.nav_upload);
            }

        }

    }

    private void openProfileFragment(View headerView) {

        ProfileFragment profileFragment = new ProfileFragment();
        Bundle bundle = new Bundle();
        bundle.putString("USER_ID", user.getId());
        bundle.putString("CURRENT_USER_ID", user.getId()); //Two same only in this case
        Log.d(TAG, "openProfileFragment: " + user.getId());
        profileFragment.setArguments(bundle);

        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.content_home, profileFragment).addToBackStack(null);
        fragmentTransaction.commit();

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.home, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        if (id == R.id.action_logout) {
            logout();
            return true;
        }

        if (id == R.id.action_blog) {
            Intent i = new Intent(Intent.ACTION_VIEW,
                    Uri.parse("http://theandroiddev.com/"));
            startActivity(i);
        }

        return super.onOptionsItemSelected(item);
    }

    void logout() {

        mFirebaseAuth.signOut();

    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.

        int id = item.getItemId();

        displaySelectedScreen(id);

        return true;
    }

    private void displaySelectedScreen(int id) {

        Fragment fragment = null;
        Bundle bundle;

        switch (id) {
            case R.id.nav_home:
                fragment = new HomeFragment();
                bundle = new Bundle();
                bundle.putString("CURRENT_USER_ID", user.getId());
                bundle.putString("USER_ID", user.getId());
                fragment.setArguments(bundle);

                break;
            case R.id.nav_ranking:
                bundle = new Bundle();
                bundle.putString("CURRENT_USER_ID", user.getId());
                bundle.putString("USER_ID", user.getId());
                fragment = new RankingFragment();
                fragment.setArguments(bundle);
                break;
            case R.id.nav_videos:
                bundle = new Bundle();
                bundle.putString("CURRENT_USER_ID", user.getId());
                bundle.putString("USER_ID", user.getId());
                fragment = new ProfileVideosFragment();
                fragment.setArguments(bundle);
                break;
            case R.id.nav_upload:
                bundle = new Bundle();
                bundle.putString("URL", urlLink);
                fragment = new InsertThreadFragment();
                fragment.setArguments(bundle);
                break;
            case R.id.nav_share:
                Toast.makeText(this, "Not ready yet :(", Toast.LENGTH_SHORT).show();
                break;
            case R.id.nav_settings:
                Toast.makeText(this, "Not ready yet :(", Toast.LENGTH_SHORT).show();
                break;

        }

        if (fragment != null) {
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.content_home, fragment).addToBackStack("stack1").commit();
        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);

    }

    @Override
    protected void onPostResume() {
        super.onPostResume();

        loadQueued();

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    public void loadQueued() {
        List<Thread> threadsInQueue = loadSharedPreferencesLogList(this);

        //Toast.makeText(this, "on RESUME" + threadsInQueue.toString(), Toast.LENGTH_SHORT).show();
        if (!threadsInQueue.isEmpty() && userIsConnected()) {

            //HomeActivity homeActivity = (HomeActivity) getActivity();

            for (int i = 0; i < threadsInQueue.size(); i++) {
                //Toast.makeText(this, "adding", Toast.LENGTH_SHORT).show();
                mDatabase.child("threads").push().setValue(threadsInQueue.get(i));
                threadsInQueue.remove(i);
            }

            saveSharedPreferencesLogList(this, threadsInQueue);
        }
    }

    private boolean userIsConnected() {

        return Utility.isNetworkAvailable(this);
//        String textt = "";
//
//        if(mFirebaseAuth.isC() != null) {
//            textt = mFirebaseAuth.getCurrentUser().getEmail();
//        }
//

    }

    @Override
    public void onAttachFragment(Fragment fragment) {
        super.onAttachFragment(fragment);
        int id = R.id.nav_home;
        if (fragment instanceof HomeFragment) id = R.id.nav_home;
        else if (fragment instanceof RankingFragment) id = R.id.nav_ranking;
        else if (fragment instanceof VideosFragment) id = R.id.nav_videos;
        else if (fragment instanceof UploadFragment) id = R.id.nav_upload;
        else if (fragment instanceof ProfileFragment) id = R.id.nav_profile_img;
        else return;
        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.getHeaderView(id);
        navigationView.setCheckedItem(id);
    }

    public void setName() {

        navNameTv.setText(user.getName());
        navEmailTv.setText(user.getEmail());
        Picasso.with(this).load(user.getPhotoUrl()).into(navProfileImg);

    }

    public void initUser() {

        if (mFirebaseAuth.getCurrentUser() != null) {
            user = new User();
            user.setId(mFirebaseAuth.getCurrentUser().getUid());
            user.setName(mFirebaseAuth.getCurrentUser().getDisplayName());
            user.setEmail(mFirebaseAuth.getCurrentUser().getEmail());
            user.setPhotoUrl(mFirebaseAuth.getCurrentUser().getPhotoUrl().toString());
        }
    }

    @Override
    public void onFragmentInteraction(Uri uri) {

    }

}

File: app/src/main/java/com/theandroiddev/riffking/HomeActivity.kt
package com.theandroiddev.riffking

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_home.*
import kotlinx.android.synthetic.main.nav_header_home.*

class HomeActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener, HomeFragment.OnFragmentInteractionListener, RankingFragment.OnFragmentInteractionListener, ThreadFragment.OnFragmentInteractionListener, UploadFragment.OnFragmentInteractionListener, ProfileFragment.OnFragmentInteractionListener, ProfileCommentsFragment.OnFragmentInteractionListener, ProfileVideosFragment.OnFragmentInteractionListener, ProfileRepFragment.OnFragmentInteractionListener, ProfileRepFragmentMe.OnFragmentInteractionListener {

    var fab: FloatingActionButton? = null

    var urlLink: String? = ""
    private var firebaseAuth = FirebaseAuth.getInstance()
    private var authStateListener: FirebaseAuth.AuthStateListener
    private var database: DatabaseReference = FirebaseDatabase.getInstance().reference

    override fun onStart() {
        super.onStart()
        firebaseAuth.addAuthStateListener(authStateListener)
    }

    init {

        authStateListener = FirebaseAuth.AuthStateListener { firebaseAuth ->
            if (firebaseAuth.currentUser == null) {
                startActivity(Intent(this@HomeActivity, LoginActivity::class.java))
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)

        //        checkGuestMode();


        initUser()

        database.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val userId = user?.id
                if (userId != null) {
                    if (dataSnapshot.child("users").child(userId).getValue(User::class.java) == null) {
                        logout()
                    }
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })

        authStateListener = FirebaseAuth.AuthStateListener { firebaseAuth ->
            if (firebaseAuth.currentUser == null) {
                startActivity(Intent(this@HomeActivity, LoginActivity::class.java))
            }
        }


        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        fab = findViewById<View>(R.id.fab) as FloatingActionButton

        val toggle = ActionBarDrawerToggle(
                this, drawer_layout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer_layout.setDrawerListener(toggle)
        toggle.syncState()

        nav_view.setNavigationItemSelectedListener(this)
        val headerView = nav_view.getHeaderView(0)
        nav_profile_img.setOnClickListener {
            openProfileFragment(headerView)
            drawer_layout.closeDrawer(GravityCompat.START)
        }

        setName()

        val fragment = HomeFragment()
        supportFragmentManager.beginTransaction().replace(R.id.content_home, fragment).commit()

        youtubeShareable()

        //        checkGuestModeViews();

    }

    //    private void checkGuestModeViews() {
    //        if (guestMode) {
    //
    //        }
    //    }
    //
    //    private void checkGuestMode() {
    //        Intent getIntent = getIntent();
    //        guestMode = getIntent.getBooleanExtra("guestmode", false);
    //    }


    //    @Override
    //    public void onBackPressed() {
    //
    //        Log.d(TAG, "onBackPressed: " +getSupportFragmentManager().getBackStackEntryCount() );
    //        if (getSupportFragmentManager().getBackStackEntryCount() >0) {
    //            getSupportFragmentManager().popBackStack();
    //        } else {
    //            Log.d(TAG, "onBackPressed: finishing");
    //            finish();
    //        }
    //    }

    private fun youtubeShareable() {
        val extras = intent.extras
        if (extras != null) {
            urlLink = extras.getString(Intent.EXTRA_TEXT)

            if (urlLink != null && urlLink != "") {
                displaySelectedScreen(R.id.nav_upload)
                //TODO send link to nav_upload and display in TV
            }

        }

    }

    private fun openProfileFragment(headerView: View) {

        val profileFragment = ProfileFragment()
        val bundle = Bundle()
        bundle.putString("USER_ID", user?.id)
        bundle.putString("CURRENT_USER_ID", user?.id) //Two same only in this case
        Log.d(TAG, "openProfileFragment: " + user?.id)
        profileFragment.arguments = bundle

        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.content_home, profileFragment).addToBackStack(null)
        fragmentTransaction.commit()

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.home, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId


        if (id == R.id.action_settings) {
            return true
        }

        if (id == R.id.action_logout) {
            logout()
            return true
        }

        if (id == R.id.action_blog) {
            val i = Intent(Intent.ACTION_VIEW,
                    Uri.parse("http://theandroiddev.com/"))
            startActivity(i)
        }

        return super.onOptionsItemSelected(item)
    }

    internal fun logout() {

        firebaseAuth.signOut()

    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.

        val id = item.itemId

        displaySelectedScreen(id)

        return true
    }

    private fun displaySelectedScreen(id: Int) {

        var fragment: Fragment? = null
        val bundle: Bundle

        when (id) {
            R.id.nav_home -> {
                fragment = HomeFragment()
                bundle = Bundle()
                bundle.putString("CURRENT_USER_ID", user?.id)
                bundle.putString("USER_ID", user?.id)
                fragment.arguments = bundle
            }
            R.id.nav_ranking -> {
                bundle = Bundle()
                bundle.putString("CURRENT_USER_ID", user?.id)
                bundle.putString("USER_ID", user?.id)
                fragment = RankingFragment()
                fragment.arguments = bundle
            }
            R.id.nav_videos -> {
                bundle = Bundle()
                bundle.putString("CURRENT_USER_ID", user?.id)
                bundle.putString("USER_ID", user?.id)
                fragment = ProfileVideosFragment()
                fragment.arguments = bundle
            }
            R.id.nav_upload -> {
                bundle = Bundle()
                bundle.putString("URL", urlLink)
                fragment = UploadFragment()
                fragment.arguments = bundle
            }
            R.id.nav_share -> share()
            R.id.nav_settings -> Toast.makeText(this, "Not ready yet :(", Toast.LENGTH_SHORT).show()
        }

        if (fragment != null) {
            supportFragmentManager.beginTransaction()
                    .replace(R.id.content_home, fragment).addToBackStack("stack1").commit()
        }

        drawer_layout.closeDrawer(GravityCompat.START)

    }

    private fun share() {
        Toast.makeText(this, "Not ready yet :(", Toast.LENGTH_SHORT).show()


    }

    override fun onPostResume() {
        super.onPostResume()

        loadQueued()

    }

    override fun onDestroy() {
        Log.d(TAG, "onDestroy: DESTROOYED")
        super.onDestroy()
    }

    fun loadQueued() {
        val threadsInQueue = loadSharedPreferencesLogList(this)

        //Toast.makeText(this, "on RESUME" + threadsInQueue.toString(), Toast.LENGTH_SHORT).show();
        if (!threadsInQueue.isEmpty() && userIsConnected()) {

            //HomeActivity homeActivity = (HomeActivity) getActivity();

            for (i in threadsInQueue.indices) {
                //Toast.makeText(this, "adding", Toast.LENGTH_SHORT).show();
                database.child("threads").push().setValue(threadsInQueue[i])
                threadsInQueue.removeAt(i)
            }

            saveSharedPreferencesLogList(this, threadsInQueue)
        }
    }

    private fun userIsConnected(): Boolean {

        return Utility.isNetworkAvailable(this)
        //        String textt = "";
        //
        //        if(firebaseAuth.isC() != null) {
        //            textt = firebaseAuth.getCurrentUser().getEmail();
        //        }
        //

    }

    //TODO HANDLE SAVING FRAGMENTS WHEN ACTIVITY IS RESTORED
    override fun onAttachFragment(fragment: Fragment?) {
        super.onAttachFragment(fragment)
        val id: Int
        when (fragment) {
            is HomeFragment -> id = R.id.nav_home
            is RankingFragment -> id = R.id.nav_ranking
            is ProfileVideosFragment -> id = R.id.nav_videos
            is ProfileFragment -> id = R.id.nav_profile_img
            is UploadFragment -> id = R.id.nav_upload
            else -> return
        }
        nav_view.getHeaderView(id)
        nav_view.setCheckedItem(id)
    }

    fun setName() {
        if (guestMode == false) {
            nav_name_tv.text = user?.name
            nav_email_tv.text = user?.email
            Picasso.get().load(user?.photoUrl).into(nav_profile_img)
        } else {
            nav_name_tv.text = "Guest"
        }

    }

    fun initUser() {

        if (firebaseAuth.currentUser != null) {
            user = User()
            user?.id = firebaseAuth.currentUser?.uid
            user?.name = firebaseAuth.currentUser?.displayName
            user?.email = firebaseAuth.currentUser?.email
            user?.photoUrl = firebaseAuth.currentUser?.photoUrl.toString()
        }
    }

    override fun onFragmentInteraction(uri: Uri) {

    }

    companion object {

        private val TAG = "HomeActivity"
        var guestMode: Boolean? = false
        var user: User? = null

        fun loadSharedPreferencesLogList(context: Context): MutableList<Thread> {
            val mPrefs = context.getSharedPreferences("threadsSaved", Context.MODE_PRIVATE)
            val gson = Gson()
            val json = mPrefs.getString("myJson", "")
            return if (json.isEmpty()) {
                mutableListOf()
            } else {
                val type = object : TypeToken<List<Thread>>() {

                }.type
                gson.fromJson(json, type)
            }
        }

        fun saveSharedPreferencesLogList(context: Context, callLog: List<Thread>) {
            val mPrefs = context.getSharedPreferences("threadsSaved", Context.MODE_PRIVATE)
            val prefsEditor = mPrefs.edit()
            val gson = Gson()
            val json = gson.toJson(callLog)
            prefsEditor.putString("myJson", json)
            prefsEditor.commit()
        }
    }

}
